runtime/vm/UpcallVMHelpers.cpp

```diff
@@ -240,24 +237,52 @@ native2InterpJavaUpcallImpl(J9UpcallMetaData *data, void *argsListPointer)
 	}
 
 	VM_VMAccess::inlineEnterVMFromJNI(currentThread);
-	returnType = getReturnTypeFromMetaData(data);
-	returnsObject = (J9NtcPointer == returnType) || (J9NtcStruct == returnType);
 
-	if (buildCallInStackFrameHelper(currentThread, &newELS, returnsObject)) {
+	for (I_32 argIndex = 0; argIndex < paramCount; argIndex++) {
+		U_8 argSigType = sigArray[argIndex].type & J9_FFI_UPCALL_SIG_TYPE_MASK;
+		j9object_t memArgObject = NULL;
+		j9object_t nativeArgArray = NULL;
+
+		if ((J9_FFI_UPCALL_SIG_TYPE_POINTER == argSigType)
+			|| (J9_FFI_UPCALL_SIG_TYPE_STRUCT == argSigType)
+		) {
+			if (J9_FFI_UPCALL_SIG_TYPE_POINTER == argSigType) {
+				I_64 offset = *(I_64*)getArgPointer(nativeSig, argsListPointer, argIndex);
+				memArgObject = createMemAddressObject(data, offset);
```

Comment:
You can't allocate objects at this point - the call-in frame has to be there first.

Reply:
This may mean this whole idea is wrong though I agree the existing code is incorrect.

Reply:
I'll need to think about this (and look in the stack walker).

Reply:
I've looked at the stack walker, and I think you're correct.

I assume the complexity here stems from the fact that you're building up the arguments for a method call, some of which are newly-allocated objects and some of which are scalars. The stack walker certainly can not handle this case, so your solution seems good.

To test this, please temporarily add an explicit call to the GC after the object allocations in this code, which should prove that the stack walker is operating correctly.

Reply:
Many thanks for double-checking the stack walker and confirming this solution.

Reply:
Just added the local GC call out of the loop after the storing the java array.

Reply:
I move the loop code for allocating objects to a new function called `storeMemArgObjectsToJavaArray` where I place the GC call before each struct/pointer object is allocated

Reply:
@gacholio, can we safely remove these explicit GC calls now given there is nothing wrong detected in tests?

Reply:
Yes.

Reply:
Just removed these two explicit GC calls in `storeMemArgObjectsToJavaArray`.